File: MVPJavaRx/app/src/main/java/com/vascome/fogtail/ui/carousel/CarouselAppFragment.java
package com.vascome.fogtail.ui.carousel;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vascome.fogtail.FogtailApplication;
import com.vascome.fogtail.R;
import com.vascome.fogtail.api.entities.RecAreaItem;
import com.vascome.fogtail.databinding.RecyclerRefreshableViewFragmentBinding;
import com.vascome.fogtail.models.AnalyticsModel;
import com.vascome.fogtail.models.AppImageLoader;
import com.vascome.fogtail.models.RecAreaItemsModel;
import com.vascome.fogtail.ui.base.fragments.BaseFragment;
import com.vascome.fogtail.ui.collectionbase.CollectionAreaItemListener;
import com.vascome.fogtail.ui.collectionbase.CollectionPresenter;
import com.vascome.fogtail.ui.collectionbase.ICollectionView;
import com.vascome.fogtail.ui.detail.RecAreaItemDetailActivity;
import com.vascome.fogtail.ui.gallery.adapter.GalleryAreaAdapter;
import com.vascome.fogtail.ui.table.decorator.BoxSpaceItemDecoration;
import com.vascome.fogtail.utils.schedulers.SchedulerProvider;

import java.util.List;

import javax.inject.Inject;

import dagger.Module;
import dagger.Provides;
import dagger.Subcomponent;


/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public class CarouselAppFragment extends BaseFragment implements ICollectionView, CollectionAreaItemListener {

    Context appContext;
    GalleryAreaAdapter galleryAreaAdapter;

    @Inject
    CollectionPresenter presenter;

    @Inject
    AppImageLoader networkBitmapClient;

    RecyclerRefreshableViewFragmentBinding binding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        appContext = getActivity().getApplicationContext();
        FogtailApplication.get(appContext).applicationComponent().plus(new CarouselFragmentModule()).inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        binding = DataBindingUtil.inflate(inflater, R.layout.recycler_refreshable_view_fragment, container, false);
        return binding.getRoot();
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initRecyclerView();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.reloadData();
    }

    @Override
    public void setLoadingIndicator(boolean active) {

        if (getView() == null) {
            return;
        }
        binding.recyclerViewSwipeRefresh.post(() -> binding.recyclerViewSwipeRefresh.setRefreshing(active));
    }

    @Override
    public void showItems(List<RecAreaItem> items) {
        runOnUiThreadIfFragmentAlive(() -> {
            setLoadingIndicator(false);
            binding.itemsLoadingErrorUi.setVisibility(View.GONE);
            binding.recyclerViewSwipeRefresh.setVisibility(View.VISIBLE);
            if (galleryAreaAdapter != null) {
                galleryAreaAdapter.setData(items);
            }
        });
    }

    @Override
    public void showError() {

        runOnUiThreadIfFragmentAlive(() -> {
            setLoadingIndicator(false);
            binding.recyclerViewSwipeRefresh.setVisibility(View.GONE);
            binding.itemsLoadingErrorUi.setVisibility(View.VISIBLE);
        });

    }

    private void initRecyclerView() {

        binding.recyclerView.addItemDecoration(new BoxSpaceItemDecoration((int) getResources().getDimension(R.dimen.list_item_vertical_space_between_items)));
        CarouselLayoutManager llm = new CarouselLayoutManager(appContext, LinearLayoutManager.HORIZONTAL, false);
        binding.recyclerView.setLayoutManager(llm);
        galleryAreaAdapter = new GalleryAreaAdapter(getActivity().getLayoutInflater(), networkBitmapClient, this);
        binding.recyclerView.setAdapter(galleryAreaAdapter);
        binding.recyclerViewSwipeRefresh.setOnRefreshListener(()->presenter.reloadData());

        presenter.bindView(this);
    }

    @Override
    public void onDestroyView() {
        presenter.unbindView(this);
        binding.unbind();
        super.onDestroyView();
    }

    @Override
    public void onItemClick(RecAreaItem clickedItem) {
        Intent intent = new Intent(appContext, RecAreaItemDetailActivity.class);
        intent.putExtra("item", clickedItem);
        startActivity(intent);
    }

    @Subcomponent(modules = CarouselFragmentModule.class)
    public interface CarouselFragmentComponent {
        void inject(@NonNull CarouselAppFragment itemsFragment);
    }

    @Module
    public class CarouselFragmentModule {

        @Provides
        @NonNull
        public CollectionPresenter provideItemsPresenter(@NonNull RecAreaItemsModel itemsModel,
                                                         @NonNull AnalyticsModel analyticsModel,
                                                         @NonNull SchedulerProvider schedulerProvider) {
            return new CollectionPresenter(itemsModel, analyticsModel,schedulerProvider);
        }
    }

}

File: MVVMKotlin/app/src/main/java/com/vascome/fogtail/presentation/main/fragment/carousel/CarouselAppFragment.kt
package com.vascome.fogtail.presentation.main.fragment.carousel

import android.annotation.SuppressLint
import android.content.Context
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.jakewharton.rxbinding2.support.v4.widget.RxSwipeRefreshLayout
import com.jakewharton.rxbinding2.view.RxView
import com.vascome.fogtail.R
import com.vascome.fogtail.data.network.AppImageLoader
import com.vascome.fogtail.data.thread.ExecutionScheduler
import com.vascome.fogtail.databinding.RecyclerRefreshableViewFragmentBinding
import com.vascome.fogtail.presentation.base.fragments.BaseFragment
import com.vascome.fogtail.presentation.main.CollectionViewModel
import com.vascome.fogtail.presentation.main.dto.RecAreaItem
import com.vascome.fogtail.presentation.main.fragment.gallery.adapter.GalleryAreaAdapter
import com.vascome.fogtail.presentation.main.fragment.table.decorator.BoxSpaceItemDecoration
import com.vascome.fogtail.presentation.main.utils.CollectionAreaItemListener
import io.reactivex.rxkotlin.addTo

import javax.inject.Inject


/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

class CarouselAppFragment : BaseFragment(), CollectionAreaItemListener {

    internal lateinit var binding: RecyclerRefreshableViewFragmentBinding
    private lateinit var galleryAreaAdapter: GalleryAreaAdapter

    @Inject
    lateinit var appContext: Context

    @Inject
    lateinit var viewModel: CollectionViewModel

    @Inject
    lateinit var imageLoader: AppImageLoader

    @Inject
    lateinit var scheduler: ExecutionScheduler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = DataBindingUtil.inflate(inflater, R.layout.recycler_refreshable_view_fragment, container, false)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initRecyclerView()
    }

    override fun onResume() {
        super.onResume()
        subscribeEvents()
        viewModel.refreshCommand.accept(true)
    }

    @SuppressLint("RxSubscribeOnError")
    private fun subscribeEvents() {
        /*
        viewModel.inProgress
                .observeOn(scheduler.UI())
                .subscribe(RxSwipeRefreshLayout.refreshing(binding.recyclerViewSwipeRefresh))
                .addTo(disposables)

        RxSwipeRefreshLayout.refreshes(binding.recyclerViewSwipeRefresh)
                .map { _ -> true }
                .subscribe { value -> viewModel.refreshCommand.accept(value) }
                .addTo(disposables)

        viewModel.items
                .observeOn(scheduler.UI())
                .subscribe({ recAreaItems ->
                    if (recAreaItems.isNotEmpty()) {
                        binding.itemsLoadingErrorUi.visibility = View.GONE
                        binding.recyclerViewSwipeRefresh.visibility = View.VISIBLE
                        galleryAreaAdapter.setData(recAreaItems)
                    } else {
                        binding.recyclerViewSwipeRefresh.visibility = View.GONE
                        binding.itemsLoadingErrorUi.visibility = View.VISIBLE
                    }
                })
                .addTo(disposables)

        RxView.clicks(binding.itemsLoadingErrorTryAgainButton)
                .subscribe { _ -> viewModel.refreshCommand.accept(true) }
                .addTo(disposables)
                */
    }

    private fun initRecyclerView() {

        binding.recyclerView.addItemDecoration(BoxSpaceItemDecoration(resources.getDimension(R.dimen.list_item_vertical_space_between_items).toInt()))
        val llm = CarouselLayoutManager(appContext, LinearLayoutManager.HORIZONTAL, false)
        binding.recyclerView.layoutManager = llm
        galleryAreaAdapter = GalleryAreaAdapter(activity.layoutInflater, imageLoader, this)
        binding.recyclerView.adapter = galleryAreaAdapter
    }

    override fun onDestroyView() {
        viewModel.destroy()
        binding.unbind()
        super.onDestroyView()
    }

    override fun onItemClick(item: RecAreaItem) {
        viewModel.openDetailCommand.accept(item)
    }
}
